Initial commit of terraform-aws-lambda

cblkwell opened the pull request:
This is moving over the module I built into a public repo so we can make it available for use by other folks. I think I've gotten the precommits and CircleCI config set up right too, but I don't know what other housekeeping we might want to include.

Commit:
Initial commit of terraform-aws-lambda

dynamike reviewed (approved):
This is awesome. I think it would be good to get @brainsik to also check review this.

brainsik commented on .circleci/config.yml:
```diff
@@ -0,0 +1,17 @@
+jobs:
+  build:
+    docker:
+      - image: trussworks/circleci-docker-primary:4a25f819d871cefc0ab71b09c2598990208a2482
```
Use latest container.
```suggestion
      - image: trussworks/circleci-docker-primary:5ca2def958c1a6abbb6ac7668c13ecee5b4598e1
```

brainsik commented on main.tf:
```diff
@@ -0,0 +1,148 @@
+ *   }
+ *
+ *   tags {
+ *     "Name" = "my_app_lambda"
```
Maybe not the best tag example? I'd expect the `Name` tag to be set automatically. Is that not the case?

cblkwell replied:
I don't think any tags get created automatically -- there's a difference between `function_name` and the *tag* Name, the same way there's a difference between the AMI `name` field and the tag Name on AMIs. Or are you saying we should automatically add that tag to every lambda? I'm wary of having too many automatically created things.

brainsik replied:
Hmm, so it sounds like for Lambda's the "Name" tag is arbitrary and not really used? If that's the case, then I still think it's not a good tag example. In AWS, the "Name" tag often is special, hence my questions.

brainsik commented on main.tf:
```diff
@@ -0,0 +1,148 @@
+
+# Create the IAM role for the Lambda instance.
+resource "aws_iam_role" "main" {
+  name               = "lambda-${var.name}-${var.job_identifier}"
```
This string gets used a lot. I think it should be a `local`.

cblkwell replied:
That works. I'll rewrite to incorporate that.

brainsik commented on main.tf:
```diff
@@ -0,0 +1,148 @@
+  name = "lambda-${var.name}-${var.job_identifier}"
+  role = "${aws_iam_role.main.id}"
+
+  policy = "${var.lambda_role_policy}"
```
It feels like the user-provided policy should be optional, in which case this inline policy should be the CloudWatch Logs policy that every lambda needs.

cblkwell replied:
I went back and forth on these a bunch...I think what I ended up picking was kind of arbitrary. No matter what I decided I felt like I was locking something in...but I think what you're saying makes perfect sense.

brainsik commented on main.tf:
```diff
@@ -0,0 +1,148 @@
+}
+
+# Attach Cloudwatch Logs policy to role defined above.
+resource "aws_iam_role_policy_attachment" "logs_policy_attach" {
```
I'm thinking that instead of creating this policy, if the user wants us to attach their own policy to the role, they should just give us the arn and they manage the creation themselves. Probably it should also be a list we take. I can imagine they have policies they reuse across lambdas and just want to pass those ARNs instead of creating new ones.

cblkwell replied:
That's a good idea; I'll rewrite to put this in.

brainsik reviewed (changes requested):
This looks amazing! Great work, Chas.

My only issue is the way the user define role policy is being handled. The setup feels backwards to me. I think the CloudWatch Logs policy shouldn't be a managed policy that's attached, but be an inline policy. It's so fundamental.

Continuing, I think the user should be able to specify zero or more policies to attach to the role. In that light, I see two things falling out:

1. It should be optional that they pass in a policy
1. We should just take a list of ARNs to attach and not manage the policy. It's the user's policy, let them manage it. I totally can see them using a single managed policy across multiple lambdas. I can also imagine a Lambda that doesn't need any additional access to AWS resources.

That's my take. Let me know if you'd like to talk it through, it's possible I'm missing something about this setup.

Commit:
Changes based on PR feedback

Commit:
Merge branch 'master' into cblkwell-initial-commit

cblkwell commented:
@brainsik I think I've addressed your concerns in the latest update of this build -- please have another look and let me know if it looks good to you.

Edit: I am bad at Git, evidently. Fixed the CircleCI issue.

Commit:
Fixing CircleCI config again

brainsik reviewed (approved):
AbFab
